drivers/led_strip/ws2812_i2s.c

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (c) 2022 Jonathan Rico
+ *
+ * Adapted from the SPI driver, using the procedure in this blog post:
+ * https://electronut.in/nrf52-i2s-ws2812/
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT worldsemi_ws2812_i2s
+
+#include <string.h>
+
+#include <zephyr/drivers/led_strip.h>
+
+#define LOG_LEVEL CONFIG_LED_STRIP_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ws2812_i2s);
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2s.h>
+#include <zephyr/dt-bindings/led/led.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+
+#define WS2812_I2S_PRE_DELAY_WORDS 1
+#define WS2812_I2S_LRCK_HZ	   100000
+
+struct ws2812_i2s_cfg {
+	struct device const *dev;
+	size_t tx_buf_bytes;
+	struct k_mem_slab *mem_slab;
+	uint8_t num_colors;
+	const uint8_t *color_mapping;
+	uint16_t reset_words;
+	bool active_low;
+	uint8_t nibble_one;
+	uint8_t nibble_zero;
+};
+
+/* Serialize an 8-bit color channel value into two 16-bit I2S words (or 1 32-bit
+ * word).
+ */
+static inline void ws2812_i2s_ser(uint32_t *word, uint8_t color, const uint8_t sym_one,
+				  const uint8_t sym_zero)
+{
+	*word = 0;
+	for (uint16_t i = 0; i < 8; i++) {
+		if ((1 << i) & color) {
+			*word |= sym_one << (i * 4);
+		} else {
+			*word |= sym_zero << (i * 4);
+		}
+	}
+
+	/* Swap the two I2S words due to the (audio) channel TX order. */
+	*word = (*word >> 16) | (*word << 16);
+}
+
+static int ws2812_strip_update_rgb(const struct device *dev, struct led_rgb *pixels,
+				   size_t num_pixels)
+{
+	const struct ws2812_i2s_cfg *cfg = dev->config;
+	uint8_t sym_one, sym_zero;
+	uint32_t reset_word;
+	uint32_t *tx_buf;
+	uint32_t flush_time_us;
+	void *mem_block;
+	int ret;
+
+	if (cfg->active_low) {
+		sym_one = (~cfg->nibble_one) & 0x0F;
+		sym_zero = (~cfg->nibble_zero) & 0x0F;
+		reset_word = 0xFFFFFFFF;
+	} else {
+		sym_one = cfg->nibble_one & 0x0F;
+		sym_zero = cfg->nibble_zero & 0x0F;
+		reset_word = 0;
+	}
+
+	/* Acquire memory for the I2S payload. */
+	ret = k_mem_slab_alloc(cfg->mem_slab, &mem_block, K_SECONDS(10));
+	if (ret < 0) {
+		LOG_ERR("Unable to allocate mem slab for TX (err %d)", ret);
+		return -ENOMEM;
+	}
+	tx_buf = (uint32_t *)mem_block;
+
+	/* Add a pre-data reset, so the first pixel isn't skipped by the strip. */
+	for (uint16_t i = 0; i < WS2812_I2S_PRE_DELAY_WORDS; i++) {
+		*tx_buf = reset_word;
+		tx_buf++;
+	}
```

Comment:
Is that because the SDOUT signal is pulled high when not transmitting (it is the case on the LPC55S69), and then this could affect the first bit ?

Is there a way to configure the I2S controller to pull down the SDOUT signal when not transmitting ?

Reply:
afaik, no. at least not for the nRF series. It's the first thing I looked at. But maybe some other peripherals do, not sure if it's a spec thing or implementation thing.